File: 백준/Gold/2504. 괄호의 값/괄호의 값.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = fs.readFileSync(filePath).toString().trim().split('');
let result = 0;

function solution(info) {
    const stack = [];
    for(let char of info) {
        if(char === '(' || char === '[') {
                stack.push(char);
        } else {
            let sum = 0;
            while(stack.length)  {
                const top = stack.pop();
                if(top === '(' && char === ')') {
                    stack.push(sum === 0 ? 2 : 2 * sum);
                    break;
                } else if(top === '[' && char === ']') {
                    stack.push(sum === 0 ? 3 : 3 * sum);
                    break;
                } else if(typeof top === 'number') {
                    sum += top;
                } else {
                    return 0;
                }
            }
            if(stack.length === 0) return 0;
        }
    }
    for(let num of stack) {
        if(typeof num !== 'number') return 0;
        result += num;
    }
    return result;
}

console.log(solution(input));

File: 백준/Gold/2504. 괄호의 값/괄호의 값.py
import sys
input = sys.stdin.read().splitlines()
brackets = list(input[0])

def solve(brackets):
    stack = []
    result = 0

    for bracket in brackets:
        if bracket == '(' or bracket == '[':
            stack.append(bracket)
        else:
            acc = 0
            if not stack:
                return 0

            isPair = False
                
            while stack:
                top = stack.pop()

                if top == '(' and bracket == ')':
                    if acc == 0:
                        stack.append(2)
                    else:
                        stack.append(acc * 2)
                    isPair = True
                    break
                elif top == '[' and bracket == ']':
                    if acc == 0:
                        stack.append(3)
                    else:
                        stack.append(acc * 3)
                    isPair = True
                    break
                elif isinstance(top, int):
                    acc += top
                else:
                    return 0
            if not isPair:
                return 0

    for num in stack:
        if isinstance(num, int):
            result += num
        else:
            return 0
            
    return result

print(solve(brackets))
